src/main/scala/org/deeplearning4j/scalphagozero/board/GameState.scala

```diff
@@ -41,84 +41,90 @@ class GameState(
-      } catch {
-        case _: Exception => println(" Illegal move attempted at: " + move.point.get.toCoords)
+    val nextBoard: GoBoard =
+      move match {
+        case Move.Play(point) =>
+          val nextBoard = this.board.clone()
+          try {
+            nextBoard.placeStone(nextPlayer, point)
+          } catch {
+            case _: Exception => println(" Illegal move attempted at: " + point.toCoords)
+          }
+          nextBoard
+        case _ => this.board
       }
-      nextBoard
-    } else {
-      this.board
-    }
-    new GameState(nextBoard, nextPlayer.other, Some(this), Some(move))
 
+    new GameState(nextBoard, nextPlayer.other, Some(this), Some(move))
   }
 
-  def isMoveSelfCapture(player: Player, move: Move): Boolean = {
-    if (!move.isPlay)
-      return false
-    this.board.isSelfCapture(player, move.point.get)
-  }
-
-  def doesMoveViolateKo(player: Player, move: Move): Boolean = {
-    if (!move.isPlay)
-      return false
-    if (!this.board.willCapture(player, move.point.get))
-      return false
-
-    val nextBoard = this.board.clone()
-    nextBoard.placeStone(player, move.point.get)
-    val nextSituation = (player.other, nextBoard.zobristHash)
-    this.allPreviousStates.contains(nextSituation)
+  def isMoveSelfCapture(player: Player, move: Move): Boolean =
+    move match {
+      case Move.Play(point) => this.board.isSelfCapture(player, point)
+      case _                => false
+    }
 
-  }
+  def doesMoveViolateKo(player: Player, move: Move): Boolean =
+    move match {
+      case Move.Play(point) if this.board.willCapture(player, point) =>
+        val nextBoard = this.board.clone()
+        nextBoard.placeStone(player, point)
+        val nextSituation = (player.other, nextBoard.zobristHash)
+        this.allPreviousStates.contains(nextSituation)
+      case _ => false
+    }
 
   def situation: (Player, GoBoard) = (nextPlayer, board)
 
-  def isValidMove(move: Move): Boolean = {
-    if (this.isOver)
-      return false
-    if (move.isPass || move.isResign)
-      return true
-    this.board.getColor(move.point.get).isEmpty && !this.isMoveSelfCapture(nextPlayer, move) &&
-    !this.doesMoveViolateKo(nextPlayer, move)
-  }
+  def isValidMove(move: Move): Boolean =
+    if (this.isOver) false
+    else {
+      move match {
+        case Move.Resign | Move.Pass => true
+        case Move.Play(point) =>
+          this.board.getColor(point).isEmpty &&
+          !this.isMoveSelfCapture(nextPlayer, move) &&
+          !this.doesMoveViolateKo(nextPlayer, move)
+      }
+    }
 
-  def isOver: Boolean = {
-    if (this.lastMove.isEmpty)
-      return false
-    if (this.lastMove.get.isResign)
-      return true
-    val secondLastMove = this.previousState.get.lastMove
-    if (secondLastMove.isEmpty)
-      return false
-    this.lastMove.get.isPass && secondLastMove.get.isPass
-  }
+  // Jules: What happen is `lastMove` is `Move.Play` ??
```

Comment:
Is it normal that the `Move.Play` isn't handled in this method ?

For now, with my code, the execution will crash if `lastMove` is `Some(Move.Play(_))`